partiql-ast/src/main/java/org/partiql/ast/ddl/AttributeConstraint.java

```diff
@@ -0,0 +1,109 @@
+package org.partiql.ast.ddl;
+
+import lombok.Builder;
+import lombok.EqualsAndHashCode;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.partiql.ast.AstNode;
+import org.partiql.ast.AstVisitor;
+import org.partiql.ast.expr.Expr;
+
+import java.util.Collection;
+import java.util.Collections;
+
+/**
+ * TODO docs, equals, hashcode
+ */
+public abstract class AttributeConstraint extends AstNode {
+
+    // NULL & NOT NULL
+
+    /**
+     * TODO docs, equals, hashcode
+     */
+    @Builder(builderClassName = "Builder")
+    @EqualsAndHashCode(callSuper = false)
+    public static class Null extends AttributeConstraint {
+        @Nullable
+        public final String name;
+
+        @NotNull
+        public final Boolean isNullable;
+
+        public Null(@Nullable String name, @NotNull Boolean isNullable) {
+            this.name = name;
+            this.isNullable = isNullable;
+        }
+
+        @NotNull
+        @Override
+        public Collection<AstNode> children() {
+            return Collections.emptyList();
+        }
+
+        @Override
+        public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
+            return visitor.visitNullable(this, ctx);
+        }
+    }
+
+    // Unique and primary
+
+    /**
+     * TODO docs, equals, hashcode
+     */
+    @Builder(builderClassName = "Builder")
+    @EqualsAndHashCode(callSuper = false)
+    public static class Unique extends AttributeConstraint {
+        @Nullable
+        public final String name;
+
+        @NotNull
+        public final Boolean isPrimary;
```

Comment:
The table one is separated because the possibility of having different target (hence not so straight forward to do constant folding). 

i.e., 

```
CREATE TABLE foo (
   attr1 INT2, 
   attr2 INT2, 
   PRIMARY KEY (attr1),
   UNIQUE (attr2)
)
```

For inline ones: 
Consider :

1. 
```
CREATE TABLE foo (
   attr1 INT2 PRIMARY KEY UNIQUE
   attr2 INT2
)
```

The same as
```
CREATE TABLE foo (
   attr1 INT2 PRIMARY KEY
   attr2 INT2
)
```

2. 
```
CREATE TABLE foo (
   attr1 INT2 PRIMARY KEY UNIQUE
   attr2 INT2
)
```

The same as
```
CREATE TABLE foo (
   attr1 INT2 PRIMARY KEY
   attr2 INT2
)
```

3. 
```
CREATE TABLE foo (
   attr1 INT2 PRIMARY KEY UNIQUE PRIMARY KEY
   attr2 INT2
)
```

Potentially illegal **or** the same as
```
CREATE TABLE foo (
   attr1 INT2 PRIMARY KEY
   attr2 INT2
)
```

So in the constraint analysis: 

1. for each node: constraint folding
2. for across column declaration: select all inline constraint that are unique -> fold to collection unique column and note inline PK. If multiple throw: 
3. for table analysis: append unique constraint columns to the unique set, make decision on valid primary key declaration based on if inline PK exists.

Reply:
Offline discussion: We want to make the Unique/Primary key constraint at attribute level and at table table level modeled more consistently. 

I will remodel the table level PK /Unique Constraint to use the boolean flag.